pkg/operator/osd/storage.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2016 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package osd
+
+import (
+	cephosd "github.com/rook/rook/pkg/cephmgr/osd"
+)
+
+type StorageSpec struct {
+	Nodes       []StorageNode    `json:"nodes,omitempty"`
+	UseAllNodes bool             `json:"useAllNodes,omitempty"`
+	Selection   StorageSelection `json:"selection,omitempty"`
+	Config      StorageConfig    `json:"config,omitempty"`
+}
+
+type StorageNode struct {
+	Name        string             `json:"name,omitempty"`
+	Devices     []StorageDevice    `json:"devices,omitempty"`
+	Directories []StorageDirectory `json:"directories,omitempty"`
+	Selection   StorageSelection   `json:"selection,omitempty"`
+	Config      StorageConfig      `json:"config,omitempty"`
+}
+
+type StorageDevice struct {
+	Name string `json:"name,omitempty"`
+}
+
+type StorageDirectory struct {
+	Path string `json:"path,omitempty"`
+}
+
+type StorageSelection struct {
+	// Whether to consume all the storage devices found on a machine
+	UseAllDevices *bool `json:"useAllDevices,omitempty"`
+
+	// A regular expression to allow more fine-grained selection of devices on nodes across the cluster
+	DeviceFilter string `json:"deviceFilter,omitempty"`
+
+	MetadataDevice string `json:"metadataDevice,omitempty"`
+}
+
+type StorageConfig struct {
+	StoreConfig  cephosd.StoreConfig `json:"storeConfig,omitempty"`
```

Comment:
Man, that's the way i first tried to implement this!  I tried to use embedded structs (composition) but could NOT get Kubernetes to marshall the cluster TPR yaml into the correct strongly typed golang struct.  If I could get that to work it would be nice because it removes a couple layers in the schema hierarchy.

I actually got a similar example working on the golang playground, but could not for the life of me get it working in Kubernetes.  I will try again though, because it makes for a cleaner cluster TPR schema.